### ISSUE TITLE
How to learn  pesudo tokens in soft prompt ?

### ISSUE THREAD

**Issue by linmou** [mayubo2333/PAIE#7]
It may be a silly question but haunts in my mind. As mentioned in the paper, pesudo tokens in soft prompts are learnable. However, I found the training process of PAIE with soft prompt is pretty similar to PAIE with manual prompts. They share the same model architecture which means, when backwarding, the loss object are from the same model. So how to learn pseudo tokens?  In the paper, you mentioned that you follow two previous method, but how to implement them in PAIE code?

**Comment by mayubo2333** [comment]
Hi, thanks for your interest on our work.

1. The soft prompt here is implemented by inserting newly pseudo tokens to vocabulary, while manual prompt uses the original word in the vocabulary. Their learning processes are indeed same.
2. We just follow their methods to construct prompts. Please refer to their official codebases for implementation.

**Comment by mayubo2333** [comment]
> It may be a silly question but haunts in my mind. As mentioned in the paper, pesudo tokens in soft prompts are learnable. However, I found the training process of PAIE with soft prompt is pretty similar to PAIE with manual prompts. They share the same model architecture which means, when backwarding, the loss object are from the same model. So how to learn pseudo tokens? In the paper, you mentioned that you follow two previous method, but how to implement them in PAIE code?

Hi, thanks for your interest on our work.

The soft prompt here is implemented by inserting newly pseudo tokens to vocabulary, while manual prompt uses the original word in the vocabulary. Their learning processes are indeed same.
We just follow their methods to construct prompts. Please refer to their official codebases for implementation.

**Comment by linmou** [comment]
Thanks for your reply.

So does that mean the only difference is the initial tokens which construct the prompts? Coz I didn't find any additional module, like an LSTM in P-tuning, that fine-tunes the prompt's embedding.

**Comment by mayubo2333** [comment]
> Thanks for your reply.
> 
> So does that mean the only difference is the initial tokens which construct the prompts? Coz I didn't find any additional module, like an LSTM in P-tuning, that fine-tunes the prompt's embedding.

Yes. And from my understanding, an additional LSTM module like P-tuning v1 is not necessary for soft prompt.
